File: library/src/main/ts/libraryBrs/src/app/core/services/rest/books.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { Book, BookCopy, BookFileColumn, BookFIlter, ChangeProposal, Opinion, Response, UserListElement } from './api/api';
import { RestService } from './rest.service';
const URL = "api/books";

@Injectable()
export class BooksService {
  public books: BehaviorSubject<[]> = new BehaviorSubject([]);
  public borrowedBooks: BehaviorSubject<Book[]> = new BehaviorSubject(new Array());

  constructor(private restService: RestService) { }

  getNewest(number:number): Observable<any> {
    return this.restService.get(`${URL}/newest?number=${number}`);
  }

  getFiltered(pageNo:number, size:number, filter:BookFIlter ): Observable<any> {
    return this.restService.get(`${URL}/filtered?pageNo=${pageNo}&size=${size}&bookFilterDto=${encodeURIComponent(JSON.stringify(filter))}`);
  }

  getPopular(number:number): Observable<any> {
    return this.restService.get(`${URL}/popular?number=${number}`);
  }
  
  getById(id:string): Observable<Response<any>> {
    return this.restService.get(`${URL}/${id}`);
  }

  getBookCopiesByBookId(id:string): Observable<Response<any>> {
    return this.restService.get(`${URL}/${id}/bookCopies`);
  }

  getAvailabilityByBookId(id:string): Observable<Response<any>> {
    return this.restService.get(`${URL}/${id}/availability`);
  }

  getOpinionsByBookId(id:string): Observable<Response<any>> {
    return this.restService.get(`${URL}/${id}/opinions`);
  }

  getOpinionsByBookIdAndUserId(userId:string, bookId:string): Observable<Response<any>> {
    return this.restService.get(`${URL}/${bookId}/user/${userId}/opinions`);
  }

  createOpinion(opinion: Opinion): Observable<Response<any>> {
    return this.restService.post(`${URL}/opinions`, opinion);
  }

  updateOpinion(opinion: Opinion): Observable<Response<any>> {
    return this.restService.put(`${URL}/opinions`, opinion);
  }

  createChangeProposal(changeProposals: ChangeProposal[]): Observable<Response<any>> {
    return this.restService.post(`${URL}/changeProposal`, changeProposals);
  }

  createUserListElement(userListElements: UserListElement): Observable<Response<any>> {
    return this.restService.post(`${URL}/userList`, userListElements);
  }

  getUserListElementByUserAndType(userId:string, type:any): Observable<any> {
    return this.restService.get(`${URL}/userList/user/${userId}?type=${type}`);
  }

  getOpinionsByUser(userId:string): Observable<any> {
    return this.restService.get(`${URL}/user/${userId}/opinions`);
  }

  deleteUserListElement(elementId:string): Observable<any> {
    return this.restService.delete(`${URL}/userList/${elementId}`);
  }

  create(book: Book): Observable<Response<any>> {
    return this.restService.post(`${URL}`, book);
  }

  createBookCopies(bookCopies: BookCopy[]): Observable<Response<any>> {
    return this.restService.post(`${URL}/bookCopies`, bookCopies);
  }

  createQRFile(bookCopiesIds:string[]): Observable<any> {
    return this.restService.getWithBody(`${URL}/bookCopies/file`,bookCopiesIds );
  }

  getChangeProposal(bookId:string): Observable<Response<any>> {
    return this.restService.get(`${URL}/${bookId}/changeProposal`);
  }

  updateChangeProposal(changeProposal:ChangeProposal): Observable<Response<any>> {
    return this.restService.patch(`${URL}/changeProposal`, changeProposal);
  }

  update(book: Book): Observable<Response<any>> {
    return this.restService.put(`${URL}`, book);
  }

  getNewChangeProposals(): Observable<any> {
    return this.restService.get(`${URL}/changeProposal/new`);
  }

  canBorrowBookCopy(bookCopyId: string, userId: string): Observable<Response<any>> {
    return this.restService.post(`${URL}/bookCopies/${bookCopyId}/users/${userId}/canBorrow`, {});
  }

  getBookByBookCopy(bookCopyId:string): Observable<any> {
    return this.restService.get(`${URL}/bookCopies/${bookCopyId}`);
  }

  getColumnsFromImport(file: any): Observable<Response<any>> {
    return this.restService.post(`${URL}/import`, file);
  }

  createColumnsMapping(bookFileColumns: BookFileColumn[]): Observable<Response<any>> {
    return this.restService.post(`${URL}/import/mapping`, bookFileColumns);
  }

  importBooks(bookFileColumns: BookFileColumn[]): Observable<Response<any>> {
    return this.restService.post(`${URL}/import/result`, bookFileColumns);
  }





 /* getFiltered(userFilter:UserFilter): Observable<any> {
    return this.restService.getWithBody(`${URL}/filtered`, userFilter);
  }*/

}

File: library/src/main/dart/mobile/reader/lib/services/booksService.dart
import 'dart:collection';

import 'package:flutter/cupertino.dart';
import 'package:reader/services/api/api.dart';
import 'package:reader/services/restService.dart';

class BooksService extends ChangeNotifier {
  final String _url = "/api/books";

  final List<Book> _books = [];

  UnmodifiableListView<Book> get items => UnmodifiableListView(_books);

  void getById(String id) {
    RestService rest = RestService();
    rest.get<dynamic>(path: "${_url}/${id}", onSuccess: onSuccessGetById);
  }

  void onSuccessGetById(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void getFiltered(int pageNo, int size) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/filtered?pageNo=${pageNo}&&size=${size}",
        onSuccess: onSuccessGetFiltered);
  }

  void onSuccessGetFiltered(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void getBookCopiesByBookId(String id) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/${id}/bookCopies",
        onSuccess: onSuccessGetBookCopiesByBookId);
  }

  void onSuccessGetBookCopiesByBookId(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void getAvailabilityByBookId(String id) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/${id}/availability",
        onSuccess: onSuccessGetBookCopiesByBookId);
  }

  void onSuccessGetAvailabilityByBookId(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void canBorrowBookCopy(String bookCopyId, String userId) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/bookCopies/${bookCopyId}/users/${userId}/canBorrow",
        onSuccess: onSuccessCanBorrowBookCopy);
  }

  void onSuccessCanBorrowBookCopy(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void getBookByBookCopy(String bookCopyId) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/bookCopies/${bookCopyId}",
        onSuccess: onSuccessGetBookByBookCopy);
  }

  void onSuccessGetBookByBookCopy(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }
}
